File: packages/gbnf-experiment/data/20260909T015142Z_f8ca89ff/ported_implementation/utils/errors/input_parse_error.py
from __future__ import annotations

from typing import TYPE_CHECKING

from .build_error_position import build_error_position
from .get_input_as_string import get_input_as_string

if TYPE_CHECKING:  # pragma: no cover
    from ...grammar_graph.types import ValidInput

__all__ = ["INPUT_PARSER_ERROR_HEADER_MESSAGE", "InputParseError"]

INPUT_PARSER_ERROR_HEADER_MESSAGE = "Failed to parse input string:"


class InputParseError(Exception):
    name = "InputParseError"

    def __init__(self, most_recent_input: "ValidInput", pos: int, previous_input: "ValidInput" = ""):
        self._most_recent_input = most_recent_input
        self._pos = pos
        self._previous_input = previous_input

        previous = get_input_as_string(previous_input)
        message = "\n".join([
            INPUT_PARSER_ERROR_HEADER_MESSAGE,
            "",
            *build_error_position(
                f"{previous}{get_input_as_string(most_recent_input)}",
                pos + len(previous),
            ),
        ])
        super().__init__(message)
        self.message = message

    @property
    def pos(self) -> int:
        return self._pos

    @property
    def most_recent_input(self) -> "ValidInput":
        return self._most_recent_input

    @property
    def previous_input(self) -> "ValidInput":
        return self._previous_input

    @property
    def src(self) -> str:
        return f"{get_input_as_string(self._previous_input)}{get_input_as_string(self._most_recent_input)}"

    @property
    def error_for_most_recent_input(self) -> str:
        return "\n".join([
            INPUT_PARSER_ERROR_HEADER_MESSAGE,
            "",
            *build_error_position(get_input_as_string(self._most_recent_input), self._pos),
        ])

    # camelCase alias, matching the reference API
    @property
    def errorForMostRecentInput(self) -> str:
        return self.error_for_most_recent_input

File: proposed-projects/round-trip-experiment/reverse/20260909T015142Z_f8ca89ff/20260910T053604Z_b86749e4/ported_implementation/src/utils/errors/input-parse-error.ts
import type { ValidInput } from '../../grammar-graph/types';
import { buildErrorPosition } from './build-error-position';
import { getInputAsString } from './get-input-as-string';

export const INPUT_PARSER_ERROR_HEADER_MESSAGE = 'Failed to parse input string:';

export class InputParseError extends Error {
  private _mostRecentInput: ValidInput;
  private _pos: number;
  private _previousInput: ValidInput;

  constructor(mostRecentInput: ValidInput, pos: number, previousInput: ValidInput = '') {
    const previous = getInputAsString(previousInput);
    super(
      [
        INPUT_PARSER_ERROR_HEADER_MESSAGE,
        '',
        ...buildErrorPosition(
          `${previous}${getInputAsString(mostRecentInput)}`,
          pos + previous.length
        ),
      ].join('\n')
    );
    this.name = 'InputParseError';
    this._mostRecentInput = mostRecentInput;
    this._pos = pos;
    this._previousInput = previousInput;
  }

  public get pos(): number {
    return this._pos;
  }

  public get mostRecentInput(): ValidInput {
    return this._mostRecentInput;
  }

  public get previousInput(): ValidInput {
    return this._previousInput;
  }

  public get src(): string {
    return `${getInputAsString(this._previousInput)}${getInputAsString(this._mostRecentInput)}`;
  }

  public get errorForMostRecentInput(): string {
    return [
      INPUT_PARSER_ERROR_HEADER_MESSAGE,
      '',
      ...buildErrorPosition(getInputAsString(this._mostRecentInput), this._pos),
    ].join('\n');
  }
}
